owb.c

```diff
@@ -108,10 +108,10 @@ static bool _reset(const OneWireBus * bus)
         _us_delay(bus->timing->G);
         gpio_set_level(bus->gpio, 0);  // Drive DQ low
         _us_delay(bus->timing->H);
-        gpio_set_level(bus->gpio, 1);  // Release the bus
+        gpio_set_direction(bus->gpio, GPIO_MODE_INPUT); // Release the bus
+        gpio_set_level(bus->gpio, 1);  // Reset the output level for the next output
```

Comment:
Does the order of `gpio_set_direction` and `gpio_set_level` matter here? Does it make sense to set the direction as input before setting the level? I don't have my DS18B20s set up at the moment however if you are getting better results from this change then perhaps it is needed.

Reply:
@DavidAntliff I wanted to release the bus as soon as possible to give the DS18B20 control so I intentionally put the direction change before setting the level. I thought about forcing the line high but the pullup should take care of that in short order. I only set the level of the gpio to 1 for consistency in case other parts of the code assumed the level was high when going back to gpio output direction.

I'm working on a blog post with the scope traces now. You can see the bus contention in the pre case and its nice and smooth in the post case (after this PR). If you give me your email address I can send them over sooner, it looks like it will be tomorrow until the post is completed.

Note that I'm still seeing those damn CRC errors but at least there isn't bus contention to cause that and all of the timings I've measured and double checked match the datasheet. Something else must be going on.

Reply:
Happy to wait for your blog post. I can try and reproduce your findings on my oscilloscope too.

How many DS18B20 devices do you have on the bus? I found with five that I needed to drop the pull-up resistor to about 2.2kOhm but I can't remember exactly why...

Reply:
@DavidAntliff only one device at the moment. I have a few other ones that I could wire up but don't really have a need for more than one device at the moment.

Reply:
@DavidAntliff alright, it took a few hours but the post is at https://chmorgan.blogspot.com/2017/11/bus-contention-on-1-wire-bus.html

I'm going to continue looking here to see if I can spot the crc issues. I'll let you know how I make out!

Reply:
Thanks for writing that up - it's very informative and clearly written. I also have a Rigol however I bought it after I wrote the library, so it's great that you discovered the bus contention. I want to double-check that the modification works with my multi-device set-up before I merge this, however I'm fairly confident it's going to be OK.

I do have a plan to modify the bus read/write code to use the RMT peripheral - this will allow the CPU to offload the timing-sensitive work to the RMT and means the library won't have to disable interrupts temporarily. This should make it a bit friendlier alongside other timing-sensitive code.

Reply:
When a CRC-verified read takes place, the device reads all 9 bytes from the scratchpad (rather than just the first two, which is where the temp data is). The last 7 bytes should be static, so maybe you could compare a failed read with a successful one and see if any of those last 7 bytes are corrupted?

Also, there was a bug in the CRC calculation that I [fixed](https://github.com/DavidAntliff/esp32-ds18b20-example/commit/9eaca8e5464d465329a1ec3559c4ddb8ca105e2c) - if you enable debug logging you should be able to see the individual steps of the CRC calculation and check that the function is converging on zero correctly (in case it's off-by-one).

Reply:
@DavidAntliff thanks, I figured I would write it up as an example of contention and screen captures are easy!

Let me know if you do find anything with the changes.

Ahh, had no idea that the RMT existed. That's cool that you can re-purpose it for that. Is that a near term plan?

I'm using the latest version of the library so I must have the crc fix in place. I am suspecting the crc but I'd like to rule out a physical issue. Should have some results coming soon on that.

Reply:
> Is that a near term plan?

It's something I plan to do soon, the problem is finding the time to do it :)

Reply:
@DavidAntliff I was able to catch what I believe to be the cause of the crc errors. The desired timing simply is not followed. There is an additional issue that the Maxim timings for the 1-wire bus are 6+9 but that the DS18B20 documentation says that the output value is only valid for 15us. Imo it seems odd to wait until the value isn't valid and then read it. Really though the DS18B20 is holding the value for what appears to be much longer, about 27us.

But in any case, I was able to catch the reads occurring far too late. In addition it is obvious that the delays are not correct. Reads are occurring pretty consistently at ~20us (the one case I caught it had occurred at 28us). So reads are late most of the time, 6+9 = 15 so there is ~5us being lost somewhere, and in at least one case I saw it ~13us late.

I found this RMT implementation and I'm going to see about integrating it into the owb library tomorrow. What do you think? https://github.com/nodemcu/nodemcu-firmware/blob/dev-esp32/components/platform/onewire.c

Reply:
That is in fact the RMT implementation I was going to use as the basis for the enhancement.

If you are thinking of giving it a go yourself, as a PR, I'd really appreciate that. If you do that, could you consider partitioning at the level of the read/write_byte functions? I.e. keep the current ones, but provide RMT-based alternatives, and use function pointers in the owb_info struct to select between implementations? This would also make the library slightly more portable (not that it really is very portable) as those functions could be substituted easily for other platforms. Happy to consider other solutions too.

I have two other repositories that contain working RMT code if you need some working examples:

https://github.com/DavidAntliff/esp32-dco
https://github.com/DavidAntliff/esp32-freqcount

Reply:
See https://github.com/DavidAntliff/esp32-owb/issues/2

Reply:
Would be good to keep all the RMT-related functions in a separate file - suggest `owb_rmt.c`.